extension/apmconfigextension/factory.go

```diff
@@ -0,0 +1,77 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package apmconfigextension // import "github.com/elastic/opentelemetry-collector-components/extension/apmconfigextension"
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/collector/component"
+	"go.opentelemetry.io/collector/extension"
+
+	"github.com/elastic/opentelemetry-collector-components/extension/apmconfigextension/apmconfig"
+	"github.com/elastic/opentelemetry-collector-components/extension/apmconfigextension/elastic/centralconfig"
+	"github.com/elastic/opentelemetry-collector-components/extension/apmconfigextension/internal/metadata"
+	"github.com/elastic/opentelemetry-lib/agentcfg"
+	"github.com/elastic/opentelemetry-lib/config/configelasticsearch"
+)
+
+func NewFactory() extension.Factory {
+	return extension.NewFactory(
+		metadata.Type,
+		createDefaultConfig,
+		createExtension,
+		metadata.ExtensionStability,
+	)
+}
+
+func createDefaultConfig() component.Config {
+	defaultElasticSearchClient := configelasticsearch.NewDefaultClientConfig()
+	return &Config{
+		AgentConfig: AgentConfig{
+			Elasticsearch: defaultElasticSearchClient,
+			// using apm-server default
+			CacheDuration: 30 * time.Second,
+		},
+		OpAMP: OpAMPConfig{
+			Server: OpAMPServerConfig{
+				Endpoint: "127.0.0.1:4320",
+			},
+		},
+	}
+}
+
+func createExtension(_ context.Context, set extension.Settings, cfg component.Config) (extension.Extension, error) {
+	extCfg := cfg.(*Config)
+	elasticsearchRemoteConfig := func(ctx context.Context, host component.Host, telemetry component.TelemetrySettings) (apmconfig.RemoteConfigClient, error) {
+		esClient, err := extCfg.AgentConfig.Elasticsearch.ToClient(ctx, host, telemetry)
+		if err != nil {
+			return nil, err
+		}
+		fetcher := agentcfg.NewElasticsearchFetcher(esClient, extCfg.AgentConfig.CacheDuration, telemetry.Logger)
+		go func() {
```

Comment:
Depending on the Go scheduler, there can be a race condition, where this anonymous Go function is never started, when `return centralconfig.NewFetcherAPMWatcher(..)` is faster.

Reply:
Would be using a sync.Waitgroup and releasing it at first inside the anonymous Go function solve this race condition?

Reply:
yes, that is the most straight forward way to make sure this go routine is created and exists.

Reply:
great! will do that. Do you know about any external source that explains this behavior? My understanding was that the `go func()` statement was already registering the routine immediately to the scheduler, even if the parent function returns earlier (in this case it is not an issue if NewFetcherAPMWatcher is scheduled first).

Reply:
I'm not aware of any guarantees by the scheduler that make sure that `go func()` is executed and does not race with the immediate `return` statement. But I had to debug and investigate such issues often enough, that I'm looking for such things and name it.

Reply:
sounds good, let's keep it safe for now https://github.com/elastic/opentelemetry-collector-components/pull/39/commits/56ba5a9a7a8841a12536c0e83d238eede9256923